00xx-account-deletion.md

```diff
@@ -0,0 +1,272 @@
+accounts* in order to propose the *zkPromise* abstraction in `o1js`.
+
+## Motivation ##
+
+Account deletion is a key protocol feature with the goal to enable the so-called
+*zkPromise* abstraction in `o1js`.
+
+Account deletion will allow to create as many accounts as you want, **then**
+receive the deposit back when you no longer need them. This is key for zkApps
+developers who want to trigger actions on other accounts in a deadlock-safe way
+so that they can create a cross-account smart contracts & applications.
+
+## Design
+
+The support for account deletion relies on adding features at the levels of
+transaction logic(s) and storage (ledgers). The latter is a prerequisite of the
+former.
+
+### Storage
+
+Within the protocol-related codebase, the lowest-level support lies in having a
+working and sound `remove` functions for the different ledger implementations
+(`Database`, `Any_ledger`, `Null_ledger`, `Syncable_ledger`) and in masking
+merkle trees.
+
+#### Merkle trees
+
+The current implementation of in-memory ledgers use a fixed-depth [Merkle
+tree](https://en.wikipedia.org/wiki/Merkle_tree).
+
+For insertion, the current implementation keeps track of the "fill frontier",
+that is, the leftmost empty slot of the tree.
+
+Now, on removal, there are (at least) 2 options:
+
+1. Symbolically mark the location as removed and make this knowledge available
+   at data structure level such that, upon new insertions, these available
+   locations are looked at before inserting on the leftmost available
+   location. This solution extends the datatype of current implementation with a
+   structure tracking freed locations.
+
+2. Switch the value currently indexed on the frontier of the tree with the
+   location that will be removed, a little bit like one does when implementing a
+   heap data structure. This solution keeps the insertion of new data to the
+   leftmost available locations and thus solution wouldn't change the current
+   datatype.
+
+Each option would correctly implement the needed support. Let us show how they
+differ.
+
+##### Option 1: Tracking freed locations
+
+To illustrate the two options, let's assume we have the following Merkle tree of depth 2, with an empty free list.
+```
+    H(H(A,B),H(C,x))
+      /       \
+   H(A,B)     H(C,x)
+   /  \       /  \
+  A    B     C    x
+```
+
+
+The insertion of data `D` results in the following Merkle tree.
+
+```
+    H(H(A,B),H(C,D))               free = []
+      /       \
+   H(A,B)     H(C,D)
+   /  \       /  \
+  A    B     C    D
+```
+
+Upon removal of `C`, the structure would evolve as follows, with `x` a
+placeholder value marking the freed location, which would actually represent to
+the empty account in practice.  The free list now states that the location
+determined by sequence of directions `[Right; Left]` is available for reuse for
+new insertions
+
+```
+    H(H(A,B),H(C,D))       free = [[Right; Left]]
```

Comment:
Indexes are already dense, shouldn't we use those?

Reply:
Do you mean using the `Bigstring` view of the same information, that is, `Merkle_address` ?

Reply:
The actual implementation will use `Location.t`, which may or may not be  a compact representation since we are within the functor's body. The current instance of this functor does use a compact representation for indexes. I chose the `Right, Left` way of wording the location because I thought it was easier to follow, especially in a tree-like drawing.